stlab/concurrency/default_executor.hpp

```diff
@@ -178,135 +221,188 @@ class task_system {
 
 class notification_queue {
     using lock_t = std::unique_lock<std::mutex>;
     std::deque<task<void()>> _q;
-    bool _done{false};
-    std::mutex _mutex;
-    std::condition_variable _ready;
+    std::mutex _queue_mutex;
+    std::condition_variable* _ready{nullptr};
 
 public:
+    void set_context(std::condition_variable& cv) {
+        _ready = &cv;
+    }
+
     bool try_pop(task<void()>& x) {
-        lock_t lock{_mutex, std::try_to_lock};
+        lock_t lock{_queue_mutex, std::try_to_lock};
         if (!lock || _q.empty()) return false;
         x = std::move(_q.front());
         _q.pop_front();
         return true;
     }
 
+    bool pop(task<void()>& x) {
+        lock_t lock{_queue_mutex};
+        if (_q.empty()) return false;
+        x = std::move(_q.front());
+        _q.pop_front();
+        return true;
+    }
+
+
     template <typename F>
     bool try_push(F&& f) {
         {
-            lock_t lock{_mutex, std::try_to_lock};
+            lock_t lock{_queue_mutex, std::try_to_lock};
             if (!lock) return false;
             _q.emplace_back(std::forward<F>(f));
         }
-        _ready.notify_one();
-        return true;
-    }
-
-    void done() {
-        {
-            lock_t lock{_mutex};
-            _done = true;
-        }
-        _ready.notify_all();
-    }
-
-    bool pop(task<void()>& x) {
-        lock_t lock{_mutex};
-        while (_q.empty() && !_done)
-            _ready.wait(lock);
-        if (_q.empty()) return false;
-        x = std::move(_q.front());
-        _q.pop_front();
+        _ready->notify_all();
         return true;
     }
 
     template <typename F>
     void push(F&& f) {
         {
-            lock_t lock{_mutex};
+            lock_t lock{_queue_mutex};
             _q.emplace_back(std::forward<F>(f));
         }
-        _ready.notify_one();
+        _ready->notify_all();
     }
 };
 
 /**************************************************************************************************/
 
-class task_system {
+class priority_task_system {
+    using lock_t = std::unique_lock<std::mutex>;
+
     const unsigned _count{std::thread::hardware_concurrency()};
-    std::vector<std::thread> _threads;
-    std::vector<notification_queue> _q{_count};
+
+    struct thread_context
+    {
+        std::mutex mutex;
+        std::condition_variable ready;
+        std::thread thread;
+    };
+    std::vector<thread_context> _thread_contexts{_count};
+    std::array<std::vector<notification_queue>, 3> _q;
     std::atomic<unsigned> _index{0};
+    std::atomic_bool _done{false};
 
     void run(unsigned i) {
         while (true) {
+            begin:
             task<void()> f;
 
-            for (unsigned n = 0; n != _count * 32; ++n) {
-                if (_q[(i + n) % _count].try_pop(f)) break;
+            for (auto& q : _q) {
+                for (unsigned n = 0; n != 64 / _count; ++n) {
```

Comment:
This is probably outside the scope of this PR, but I would be curious to see if performance can be improved by using a backoff function for the high-water mark of `n`. I toyed with this a while back on a pet project [making a hybrid spin mutex](https://hackernoon.com/building-a-c-hybrid-spin-mutex-f98de535b4ac). (Pay no attention to the article, just the delay backoff function.)

If the backoff were written properly, we could have busier thread pools grow towards a larger count for `n` before calling `pop`. Likewise, a thread pool with less activity would find itself there sooner. Just a thought.

Reply:
The basic idea of this implementation was developed by @sean-parent . He gives in his talk at Russia C++ https://www.youtube.com/watch?v=IYmoz7e1L5c details how he came to that solution.